icu4c/source/i18n/tznames_impl.cpp

```diff
@@ -731,6 +731,7 @@ class ZNames : public UMemory {
                 nameinfo->type = getTZNameType(static_cast<UTimeZoneNameTypeIndex>(i));
                 trie.put(name, nameinfo, status); // trie.put() takes ownership of the key
                 if (U_FAILURE(status)) {
+                    uprv_free(nameinfo);
```

Comment:
If trie.put() truly takes ownership, then it should do so even when something goes wrong. In that case, we would now get a double-delete problem.
Please check.
@yumaoka @aheninger FYI

Reply:
you are right. I think the real issue is the TextTrieMap::put() implement called the deleter on the wrong variable. it should call the deleter to delete the value, not the key